### ISSUE TITLE
git_push() and git_fetch() should have a default remote

### ISSUE THREAD

**Issue by jennybc** [r-lib/gert#53]
I think `git_push()` should have some default logic re: `remote`. What should it be,  though?

Docs for `git push`: <https://git-scm.com/docs/git-push>

> When the command line does not specify where to push with the `<repository>` argument, `branch.*.remote configuration` for the current branch is consulted to determine where to push. If the configuration is missing, it defaults to `origin`.

Seems like `gert::git_push()` should match that as closely as possible. The existing check against `git_info()$remote` appears to implement the check against `branch.*.remote configuration` for the current branch.

https://github.com/r-lib/gert/blob/4fe319dc8260072c46c048f5f56b04aef3791763/R/fetch.R#L50-L53

But this guard feels too aggressive:

https://github.com/r-lib/gert/blob/4fe319dc8260072c46c048f5f56b04aef3791763/R/fetch.R#L55-L56

I think this should fall back to `origin`.

I'm about to post a reprex showing how `git_push()` insists on me specifying a remote, even in a case where `git push` in the shell succeeds (falling back to `origin`).

**Comment by jennybc** [comment]
``` r
> devtools::load_all("~/rrr/gert")
Loading gert
Linking to libgit2 v0.28.4, ssh support: YES
Global config: /Users/jenny/resources/system-setup/dotfiles/DOTgitconfig
Default user: Jenny Bryan <jenny.f.bryan@gmail.com>

> r <- tempfile(pattern = "gert")
> git_init(r)
> oldwd <- setwd(r)
> writeLines(letters[1:6], "alphabet.txt")
> git_add("alphabet.txt")
# A tibble: 1 x 3
  file         status staged
  <chr>        <chr>  <lgl> 
1 alphabet.txt new    TRUE  
> git_commit("Start alphabet file")
[1] "cdd8b4f548163de384c3af53c531cfa2734a4e57"

> usethis::proj_set(getwd())
✓ Setting active project to '/private/var/folders/pv/l2mbh7j90y5_vslh_1wyqxj00000gn/T/Rtmpw6bw4U/gert152f6b2eeef8'
> usethis::use_github(private = TRUE)
✓ Checking that current branch is 'master'
● Check title and description
  Name:        gert152f6b2eeef8
  Description: 
Are title and description ok?

1: Negative
2: Nope
3: I agree

Selection: 3
✓ Creating GitHub repository
✓ Setting remote 'origin' to 'git@github.com:jennybc/gert152f6b2eeef8.git'
✓ Pushing 'master' branch to GitHub and setting remote tracking branch
✓ Opening URL 'https://github.com/jennybc/gert152f6b2eeef8'

> git_branch_create("new-branch")
> git_info()$shorthand
[1] "new-branch"
> writeLines(month.name[1:6], "months.txt")
> cat(letters[7:9], file = "alphabet.txt", sep = "\n", append = TRUE)
> git_add("alphabet.txt")
# A tibble: 2 x 3
  file         status   staged
  <chr>        <chr>    <lgl> 
1 alphabet.txt modified TRUE  
2 months.txt   new      FALSE 
> git_add("months.txt")
# A tibble: 2 x 3
  file         status   staged
  <chr>        <chr>    <lgl> 
1 alphabet.txt modified TRUE  
2 months.txt   new      TRUE  
> git_commit("Do stuff in the branch")
[1] "dbb4949e76d50b43553fd0e278ae7ef3ed98a401"

> git_branch_list()
# A tibble: 3 x 5
  name       local ref                upstream           commit                    
* <chr>      <lgl> <chr>              <chr>              <chr>                     
1 master     TRUE  refs/heads/master  refs/remotes/orig… cdd8b4f548163de384c3af53c…
2 new-branch TRUE  refs/heads/new-br… NA                 dbb4949e76d50b43553fd0e27…
3 origin/ma… FALSE refs/remotes/orig… NA                 cdd8b4f548163de384c3af53c…

> git_push()
Error in git_push() : No remote is set for this branch
Run `rlang::last_error()` to see where the error occurred.
```

But, at this point, in the shell, I can do `git push` or `git push -u` and it "just works":

``` bash
jenny@2020-mbp tmp $ cd /private/var/folders/pv/l2mbh7j90y5_vslh_1wyqxj00000gn/T/Rtmpw6bw4U/gert152f6b2eeef8

jenny@2020-mbp gert152f6b2eeef8 $ git push -u
Enumerating objects: 6, done.
Counting objects: 100% (6/6), done.
Delta compression using up to 12 threads
Compressing objects: 100% (2/2), done.
Writing objects: 100% (4/4), 368 bytes | 368.00 KiB/s, done.
Total 4 (delta 0), reused 0 (delta 0)
remote: 
remote: Create a pull request for 'new-branch' on GitHub by visiting:
remote:      https://github.com/jennybc/gert152f6b2eeef8/pull/new/new-branch
remote: 
To github.com:jennybc/gert152f6b2eeef8.git
 * [new branch]      new-branch -> new-branch
Branch 'new-branch' set up to track remote branch 'new-branch' from 'origin'.

jenny@2020-mbp gert152f6b2eeef8 $ git branch -vv
  master     cdd8b4f [origin/master] Start alphabet file
* new-branch dbb4949 [origin/new-branch] Do stuff in the branch
```

**Comment by jeroen** [comment]
So how does `git` know the remote? Does is just default to origin? Or does it inherit the remote from the original branch somehow?

**Comment by jennybc** [comment]
I now see that `git_fetch()` and `git_push()` are basically tethered together

Docs for `git fetch`: <https://git-scm.com/docs/git-fetch>

> `git fetch` can fetch from either a single named repository or URL, or from several repositories at once if `<group>` is given and there is a `remotes.<group>` entry in the configuration file. (See git-config[1]).
>
> When no remote is specified, by default the `origin` remote will be used, unless there’s an upstream branch configured for the current branch.

**Comment by jennybc** [comment]
> So how does git know the remote? Does is just default to `origin`? Or does it inherit the remote from the original branch somehow?

If the current branch has a tracking branch, it takes that remote. As gert already does. If the current branch does not have a tracking branch, it falls back to `origin`. This is my understanding now for both `git push` and `git fetch`.

*Update: as we later worked out in the PR, part of why I can do argument-less `git push` here is because I set `push.default` to `current` in my global git config. Which is not the default, but it also not an unusual or surprising choice.*

**Comment by jeroen** [comment]
Fixed via db5e6b7a8a73e249766ea56fdd9a71cab6b90a0b
